libs/libcommon/tests/test_smart_update.py

```diff
@@ -0,0 +1,283 @@
+    put_cache(step=STEP_DA, dataset=DATASET_NAME, revision=REVISION_NAME)
+    # Same revision as cache, do nothing
+    plan = get_smart_dataset_update_plan(processing_graph=PROCESSING_GRAPH_TWO_STEPS)
+    assert_smart_dataset_update_plan(plan, tasks=[])
+
+
+def test_cache_revision_is_not_parent_revision_commit() -> None:
+    # If rewriting git history OR spamming commits OR out of order commits OR in case of concurrency issues: raise
+    put_cache(step=STEP_DA, dataset=DATASET_NAME, revision="not_parent_revision")
+    with pytest.raises(SmartUpdateImpossibleBecauseCachedRevisionIsNotParentOfNewRevision):
+        get_smart_dataset_update_plan(processing_graph=PROCESSING_GRAPH_TWO_STEPS)
+
+
+def test_empty_commit() -> None:
+    # Empty commit: update the revision of the cache entries
+    put_cache(step=STEP_DA, dataset=DATASET_NAME, revision=OTHER_REVISION_NAME)
+    with put_diff(EMPTY_DIFF):
+        plan = get_smart_dataset_update_plan(processing_graph=PROCESSING_GRAPH_TWO_STEPS)
+        assert_smart_dataset_update_plan(
+            plan,
+            files_impacted_by_commit=[],
+            tasks=["UpdateRevisionOfDatasetCacheEntriesTask,1"],
+        )
+    with pytest.raises(RequestException):  # if diff doesn't exist
+        get_smart_dataset_update_plan(processing_graph=PROCESSING_GRAPH_TWO_STEPS)
+
+
+def test_add_initial_readme_commit() -> None:
+    # Add README.md commit: update the revision of the cache entries
+    put_cache(step=STEP_DA, dataset=DATASET_NAME, revision=OTHER_REVISION_NAME)
+    with put_diff(ADD_INITIAL_README_DIFF), put_readme(INITIAL_README):
+        plan = get_smart_dataset_update_plan(processing_graph=PROCESSING_GRAPH_TWO_STEPS)
+        assert_smart_dataset_update_plan(
+            plan,
+            files_impacted_by_commit=["README.md"],
+            updated_yaml_fields_in_dataset_card=[],
+            tasks=["UpdateRevisionOfDatasetCacheEntriesTask,1"],
+        )
+
+
+def test_add_data() -> None:
+    # Add data.txt commit: raise
+    put_cache(step=STEP_DA, dataset=DATASET_NAME, revision=OTHER_REVISION_NAME)
+    with put_diff(ADD_DATA_DIFF):
+        with pytest.raises(SmartUpdateImpossibleBecauseOfUpdatedFiles):
+            get_smart_dataset_update_plan(processing_graph=PROCESSING_GRAPH_TWO_STEPS)
+
+
+def test_add_config_commit() -> None:
+    # Add config commit: raise
+    put_cache(step=STEP_DA, dataset=DATASET_NAME, revision=OTHER_REVISION_NAME)
+    with put_diff(ADD_CONFIG_DIFF), put_readme(ONE_CONFIG_README):
+        with pytest.raises(SmartUpdateImpossibleBecauseOfUpdatedYAMLField):
+            get_smart_dataset_update_plan(processing_graph=PROCESSING_GRAPH_TWO_STEPS)
+
+
+def test_add_second_config_commit() -> None:
+    # Add a second config commit: raise
+    put_cache(step=STEP_DA, dataset=DATASET_NAME, revision=OTHER_REVISION_NAME)
+    with put_diff(ADD_SECOND_CONFIG_DIFF), put_readme(ONE_CONFIG_README, revision=OTHER_REVISION_NAME), put_readme(
```

Comment:
just to be sure I understand: when you call `put_readme` twice, the second one will mock the already mocked `hf_hub_download` again?

Reply:
Yes, they wrap each other